### PR TITLE
[TCFMM-31] Improve frontend functionality

### PR DESCRIPTION
**Pull request by rinn7e**
[//]: # (This is a template of a pull request template.)
[//]: # (You should modify it considering specifics of a particular repository and put it there.)
[//]: # (Comments like this are meta-comments, they shouldn't be present in the final template.)
[//]: # (Comments starting with '

[//]: # (Here you can add a link to the corresponding issue tracker, e. g. https://issues.serokell.io/issue/AD-)
[//]: # (For GitHub/GitLab issues it is better to use just hash symbol or exclamation mark as it is more resistant to repo movements)
[//]: # (In this case please also prefix the link with "Resolves" keyword)
[//]: # (See https://docs.gitlab.com/ee/user/project/issues/managing_issues.html#closing-issues-automatically)
[//]: # (See https://help.github.com/en/github/managing-your-work-on-github/closing-issues-using-keywords)
## Related issue(s)

Resolves TCFMM-31

#### Related changes (conditional)

- Tests
  - [x] If I added new functionality, I added tests covering it.
  - [x] If I fixed a bug, I added a regression test to prevent the bug from
        silently reappearing again.

[//]: # (Add more docs here if you have them in the repository)
- Documentation
  - [x] I checked whether I should update the docs and did so if necessary:
    - [Specification](../tree/master/docs/specification.md)
    - [README](../tree/master/README.md)
    - Haddock

#### Stylistic guide (mandatory)

[//]: # (Update link to style guide if necesary or remove if it's not present)

- [x] My commits comply with [the policy used in Serokell](https://www.notion.so/serokell/Where-and-how-to-commit-your-work-58f8973a4b3142c8abbd2e6fd5b3a08e).
- [x] My code complies with the [style guide](../tree/master/docs/code-style.md).

### CODE REVIEW

File: frontend/src/App.tsx [hunk 1]
```diff
@@ -124,14 +127,34 @@ export function update(msg: Msg, model: Model): [Model, cmd.Cmd<Msg>] {
...
+      console.log("what", msg.result)
+      let getStorageResult = pipe(msg.result, E.fold(RD.failure, r => RD.success(r)))
+
+      return [
+        {
+          ...model,
+          storage: getStorageResult
         }
         , cmd.none
       ]
```
Comment (pasqu4le) [inline frontend/src/App.tsx:160]: It seems that I'm unable to connect, I believe (from your question on Slack) that the hardcoded URL a few lines below should be `https://granada.testnet.tezos.serokell.team/`.

Also, I think both of these URL should be settable, I'd like to see them as options in the make target too.

File: frontend/src/Data/Form.ts [hunk 2]
```diff
@@ -72,13 +73,13 @@ export const xToYFormInit: () => Map<string, FormItem> = () =>
       ["deadline",
         {
           order: 3,
-          label: "Deadline: (example: 2022-01-01T00:01:40Z)",
+          label: "Deadline:",
```
Comment (pasqu4le) [inline frontend/src/Data/Form.ts:76]: The date picker is nice, but I think it would still be nice to give a suggestion, maybe a few minutes in the future from "now" at any given moment?

File: frontend/src/App.tsx [hunk 3]
```diff
@@ -334,42 +367,67 @@ const Form = (props: { model: Model, dispatch: React.Dispatch<Msg> }) => {
 export const EntrypointForms = (props: { model: Model, dispatch: React.Dispatch<Msg> }) => {
 
   const isDisabled = (): string => {
-    if (props.model.contract._tag === "RemoteSuccess") return 'opacity-100 pointer-events-auto'
-    else return 'opacity-50 pointer-events-none'
+    if (props.model.contract._tag === "RemoteSuccess") return ' opacity-100 pointer-events-auto'
+    else return ' opacity-50 pointer-events-none'
   }
```
Comment (pasqu4le) [inline frontend/src/App.tsx:373]: I'm not sure why, but now on any successful call that I execute I get a blank page, have you notice this?

File: frontend/src/App.tsx [hunk 4]
```diff
@@ -334,42 +367,67 @@ const Form = (props: { model: Model, dispatch: React.Dispatch<Msg> }) => {
...
+        <p className="text-sm text-green-600">{JSON.stringify(props.model.callEntrypointResult.value)}</p>
+      </div>
     else return <div></div>
   }
 
-  return <div className={"bg-white rounded-2xl mt-10 m-auto " + isDisabled()} style={{ width: "500px" }}>
+  let ContractStorageView = () => {
+    if (props.model.storage._tag === "RemoteSuccess")
+      return <div>
+        <p className="pb-1">Sqrt price: {props.model.storage.value.sqrt_price.toString()}</p>
```
Comment (pasqu4le) [inline frontend/src/App.tsx:400]: I think it would be better to show the price rather than its square root here.

File: frontend/src/App.tsx [hunk 5]
```diff
@@ -334,42 +367,67 @@ const Form = (props: { model: Model, dispatch: React.Dispatch<Msg> }) => {
...
+      return <div>Loading...</div>
+    else return <div></div>
+  }
 
-    <div className="relative">
-      <div className="absolute top-0 bg-white py-4 rounded-lg" style={{ "left": "-260px" }}>
-        {pipe(allEntrypoints, A.mapWithIndex((i, a) => <FormSelect key={i} a={a} dispatch={props.dispatch} model={props.model} />))}
+  return <div>
+    <div className={"bg-white rounded-2xl mt-10 m-auto p-6 " + isDisabled()} style={{ width: "500px" }}>
+      <p className="font-bold pb-1">Relevant Contract Storage</p>
```
Comment (pasqu4le) [inline frontend/src/App.tsx:409]: This seems a bit of an odd name, I don't think users care about the where that info came from.

I'd rephrase this as  "current contract info".

Also, what do you think about adding a button to manually refresh this?

File: frontend/src/Data/Form.ts [hunk 6]
```diff
@@ -445,3 +422,95 @@ export const increaseObservationCountFormInit: () => Map<string, FormItem> = ()
...
   )
+
+
+export const updatePositionFormInit: () => Map<string, FormItem> = () =>
+  new Map(
+    [
+      ["position_id",
+        {
+          order: 1,
+          label: "Position ID",
```
Comment (pasqu4le) [inline frontend/src/Data/Form.ts:433]: it appears that if I input a negative number for the position ID the UI freezes.

(I don't mean in any specific place, it seems to be common to all such fields)

File: frontend/README.md [hunk 7]
```diff
@@ -11,6 +11,7 @@ To communicate with the webapp, the user should install [Temple wallet](https://
 
 
 ```bash
+make typescript # To generated types needed by the frontend
```
Comment (pasqu4le) [inline frontend/README.md:14]: since we are suggesting to run `make` here, why not just suggest to run `make frontend`?

File: Makefile [hunk 8]
```diff
@@ -145,7 +145,7 @@ typescript: prepare_lib
 	# Copy generated files to the frontend
 	mkdir -p frontend/src/Generated
 	cp $(TS_OUT)/segmented-cfmm/src/common.ts frontend/src/Generated/common.ts
-	cp -r $(TS_OUT)/segmented-cfmm/src/generated/ frontend/src/Generated/generated/
+	cp -a $(TS_OUT)/segmented-cfmm/src/generated/. frontend/src/Generated/generated/
```
Comment (pasqu4le) [inline Makefile:148]: why this change?

File: Makefile [hunk 9]
```diff
@@ -150,10 +150,21 @@ typescript: prepare_lib
...
+frontend: typescript
+	mkdir -p frontend/src/Generated
+
+	# Don't print error message if the file does not exist.
+	rm frontend/src/Generated/Env.ts 2> /dev/null || true
+
+	echo "export const env = \"$(env)\"" >> frontend/src/Generated/Env.ts
+	echo "export const testnetLink = \"$(testnet_link)\"" >> frontend/src/Generated/Env.ts
+	echo "export const mainnetLink = \"$(mainnet_link)\"" >> frontend/src/Generated/Env.ts
 	cd frontend && yarn && yarn tailwind && yarn dev
```
Comment (dcastro) [inline Makefile:169]: I'm having a weird bug now...

If I select the "3 minutes" option, the contract will fail with error 103 (aka the `past_deadline_err` error).
If I manually select a time 3 minutes in the future from the datetime picker, the call works.

I recorded a short video below.

Even selecting the "1 day" option fails sometimes as well.

https://user-images.githubusercontent.com/718166/139691479-6245f9b4-24b7-42ab-886b-89289188d1cf.mp4

Comment (rinn7e) [reply Makefile:169]: That's weird, could you give me some detail on what entrypoint and parameters? Seems like the video is not working for me, with the error the file is corrupt.

Comment (dcastro) [reply Makefile:169]: Just called the `swap x for y` entrypoint, with `x=10`, but I think this happens in other entrypoints too.

Here, I tried converting the video to another format, does this one work for you?

https://user-images.githubusercontent.com/718166/139812407-e1e15255-67b0-4676-a132-aec5c977acbc.mov

File: frontend/README.md [hunk 10]
```diff
@@ -11,12 +11,27 @@ To communicate with the webapp, the user should install [Temple wallet](https://
...
+  env=development \
+  testnet_link=https://granada.testnet.tezos.serokell.team/ \
+  mainnet_link=https://mainnet.api.tez.ie/
 ```
 
 ## Limitation
 
-The web app does not support all the entrypoints, particularly the entrypoints that are needed to be called by a contract.
\ No newline at end of file
+The web app does not support all the entrypoints, particularly the entrypoints that are needed to be called by a contract.
```
Comment (pasqu4le) [inline frontend/README.md:23]:
```suggestion
The web app does not support all the entrypoints, particularly the entrypoints that need to be called by a contract.
```
sounds a bit better to me

File: Makefile [hunk 11]
```diff
@@ -150,10 +150,21 @@ typescript: prepare_lib
...
 	mkdir -p frontend/src/Generated
 	cp $(TS_OUT)/segmented-cfmm/src/common.ts frontend/src/Generated/common.ts
-	cp -r $(TS_OUT)/segmented-cfmm/src/generated/ frontend/src/Generated/generated/
+	cp -a $(TS_OUT)/segmented-cfmm/src/generated frontend/src/Generated/
 
 # Quickly setup the frontend and run the it locally. Check `frontend/readme.md` for details.
-frontend:
+frontend: env = development
+frontend: testnet_link = https://granada.testnet.tezos.serokell.team/
+frontend: mainnet_link = https://mainnet.api.tez.ie/
```
Comment (pasqu4le) [inline Makefile:158]: These should probably be named:
```suggestion
frontend: testnet_url = https://granada.testnet.tezos.serokell.team/
frontend: mainnet_url = https://mainnet.api.tez.ie/
```
note: if you agree please update them in the README too

**Review by pasqu4le — APPROVED** [review]
Some minor nitpicks, but the rest LGTM

**Review by dcastro — APPROVED** [review]
